### PR TITLE
Use coroutines for copy to s3

### PR DESCRIPTION
**Pull request by vitlibar**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
Almost green tests :rocket:

File: src/Common/CoTask.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
+#include <Common/CoTask.h>
+
+
+namespace DB::ErrorCodes
+{
+    extern const int TASK_CANCELLED;
+}
+
+
+namespace DB::Co
```
Comment (alexey-milovidov) [inline src/Common/CoTask.cpp:10]: This is too cool. Let's name it in a boring way: `Coroutine`.

Comment (vitlibar) [reply src/Common/CoTask.cpp:10]: This prefix is used by C++ itself, see the keywords `co_await`, `co_return`.

File: src/Common/CoTask.h [hunk 2]
```diff
@@ -0,0 +1,621 @@
...
+            {
+                if (cancel_exception)
+                    std::rethrow_exception(cancel_exception);
+            }
+        };
+        return Getter{{}, getAwaitControl()->getCancelStatus()->getCancelException()};
+    }
+
+    template <typename AwaiterType>
+    auto await_transform(AwaiterType && awaiter, typename std::enable_if_t<details::is_awaiter_type_v<std::remove_cvref_t<AwaiterType>>> * = nullptr)
```
Comment (serxa) [inline src/Common/CoTask.h:410]: Concepts would be simpler and not dependent on boost
https://godbolt.org/z/6b5x3Wo5o
```suggestion
    requires requires(F f) { {f.await_ready()} -> std::same_as<bool>; }
    auto await_transform(AwaiterType && awaiter)
```

**Comment by alexey-milovidov** [comment]
@zhanglistar, This solution was complex and not finished.
Although coroutines could be useful here, threads are also ok.

**Pull request closed without merging**